### ISSUE TITLE
Can't search?

### ISSUE THREAD

**Issue by elad661** [fedora-infra/bodhi#229]
I'm trying to use the search functionality to search for a package. All I get is "Unable to find any users matching the current query".

**Comment by jwakely** [comment]
You need to search for NVR not just name, e.g. foobar-1.2-3.fc23

**Comment by elad661** [comment]
That's a regression from the previous major version, where I could search the name without the version or release.

**Comment by ralphbean** [comment]
You can still search for the name without the version or release, there's just no feedback that it is still searching.  We need to add ``spinner.gif`` in there so people know to wait for results:  ![](https://bodhi.fedoraproject.org/static/img/spinner.gif)

**Comment by ralphbean** [comment]
A note on implementation for the spinner.gif thing.  I looked into it tonight, and there's no way to hook into the current version of typeahead.js we're using to **add** the spinner when we start the ajax call, but there is a hook to remove it at the end when results are returned.

The latest version of typeahead.js boasts callbacks at both the start and end.  Employing that would look something like this:

```diff
diff --git a/bodhi/static/js/search.js b/bodhi/static/js/search.js
index a1120fc..f5f7644 100644
--- a/bodhi/static/js/search.js
+++ b/bodhi/static/js/search.js
@@ -5,7 +5,16 @@ $(document).ready(function() {
         queryTokenizer: Bloodhound.tokenizers.whitespace,
         remote: {
             url: 'updates/?like=%QUERY',
-            filter: function(response) { return response.updates; },
+            prepare: function(settings) {
+                console.log('in prepare for updates!');
+                // TODO -- add the spinner.gif to the DOM here
+                return settings;
+            },
+            transform: function(response) {
+                console.log('in transform for updates!');
+                // TODO -- remove the spinner.gif to the DOM here
+                return response.updates;
+            },
         }
     });
     var users = new Bloodhound({
@@ -13,7 +22,16 @@ $(document).ready(function() {
         queryTokenizer: Bloodhound.tokenizers.whitespace,
         remote: {
             url: 'users/?like=%QUERY',
-            filter: function(response) { return response.users; },
+            prepare: function(settings) {
+                console.log('in prepare for users!');
+                // TODO -- add the spinner.gif to the DOM here
+                return settings;
+            },
+            transform: function(response) {
+                console.log('in transform for users!');
+                // TODO -- remove the spinner.gif to the DOM here
+                return response.users;
+            },
         }
     });
```

**However**, the latest version of typeahead.js breaks *everything else* about our search box and I couldn't figure it out in the short box of time I wanted to personally alot this issue.

If someone else comes along and wants to hack on this, I hope these notes will be helpful to you in figuring it out.

**Comment by ellert** [comment]
With the old bodhi searching for a package name gave you a list of all updates for that package, which was very useful. This functionality is missing in the new bodhi's search box.

You can get the list by typing the url in the browser:
https://bodhi.fedoraproject.org/updates/?packages=<packagename>

But it would be much more convenient if you could get there from the search box like in the old bodhi.

**Comment by PatrickRHinkley** [comment]
The problem is much more severe in my case than you guys realize. It's not a matter of patience for the results to load ... they take WAY too long (other than usernames, which come up instantly). I have not been able to use bodhi for the last week or so since this "upgrade" occurred.

In my testing I've figured out what's going on in my case ... by watching System Monitor, I can see that there's about a 10 second delay before downloading the results STARTS, and then it takes approximately 40 seconds on my connection for the results to download after that. Expecting some users to have to wait 50 seconds just to get a list of "suggestions" is unacceptable.

If the user presses enter without selecting from the dropdown, please do take them to https://bodhi.fedoraproject.org/updates/?packages=[query string]

**Comment by jwakely** [comment]
It's slow for me, but not 50 seconds slow, however ...

> (other than usernames, which come up instantly)

the fact that _some_ results come up instantly may be counterproductive, because it looks as though the search has finished and it just hasn't returned any builds/packages. Once you've used it a few times you remember that you need to be patient and ignore the user results and wait for the _other_ results, but that's not at all obvious when you first use the site.

**Comment by ralphbean** [comment]
Noting again here that we're limited by typeahead.js's API.  I looked into it more today and the latest version is [considered unstable](https://github.com/twitter/typeahead.js/blob/master/CHANGELOG.md) and we're advised to wait for a stable 1.0 release.

**Comment by PatrickRHinkley** [comment]
Update: because the search feature was unusably slow for me, I gave up on using the new Bodhi after a couple weeks. Tonight I came back to see if the issue has been resolved, and I can report that the suggestions now come up near instantly.

Actually, as I continue to test while posting this, I notice that the change appears to be that a "Packages" category comes up near instantly. The "Updates" category takes about 7 additional seconds to load, but I personally don't care because I just want to search for a package, plus 7 seconds is way better than the 50 second wait I was dealing with before.

The only remaining issue from my perspective is that the user is not allowed to simply press enter to start their search, instead being forced to select an entry from the "suggestions" dropdown.
